Update Event Bus architecture (#3569)
kyma-project/kyma: docs/event-bus/docs/02-01-event-bus.md

title: Architecture
---

See the diagram overview of the Event flow configuration and Event cons

> ![Configure and Consume Events](./assets/configure-consume-events.svg)

Reviewer: I would suggest starting with the same color. It seems simpler to me in that case.

## Eventing configuration 

1. A user configures a lambda or a serverless application to be triggered by an Event from an external solution. 

> >**NOTE**: For a serverless application the user must create a Kyma Subscription resource manually. For a lambda, the it is created automatically.

Reviewer: The diagram should somehow depict that it belongs to event consumption and configuration. Currently, it precedes that section

title: Architecture
---

## Event consumption

When you create a lambda or a service to perform a given business functionality, you must define which Events trigger it. Define triggers by creating the [Subscription CR](components/event-bus/#custom-resource-subscription) where you instruct the Event Bus to forward the Events of a particular type to your lambda. 

> For example, whenever the `order-created` Event comes in the Event Bus stores it in NATS Streaming. It then dispatches it to the receiver specified in the Subscription CR definition.

Author: Done.

> > **NOTE:** The Event Bus creates a separate Event Trigger for each Subscription.

Author: Done.

![Configure and Consume Events](./assets/configure-consume-events.svg)

> 1. A user creates a lambda or a service to be triggered by an Event coming from an external solution. 

Author: Done.

    >**NOTE**: For a service, the user must create a Kyma Subscription resource manually. For a lambda, it is created automatically.

> 2. The **subscription-controller-knative** component reacts on the creation of Kyma Subscription.  It [verifies](#event-validation) if the Event type from the Application can be consumed in the Namespace where the Kyma Subscription has been created.  If so, it creates the Knative Channel and Knative Subscription resources.

Author: application :)

> 3. The **nats-controller** reacts on the creation of a Knative Channel and creates the required Kubernetes and Istio services.

Author: Done.

> 4. The **nats-dispatcher** reacts on the creation of a Knative Subscription and creates the NATS Streaming subscription. 

Author: It will be Subscription.

> 5. The **nats-dispatcher** picks the Event and dispatches it to the configured lambda or the service URL as an HTTP POST request. The lambda reacts on the received Event.

Author: Done.

## Event publishing

![Publish Events](./assets/publish-events.svg)

1. The external application integrated with Kyma makes a REST API request to the Application Connector's Events Gateway to indicate that a new Event is available. The request provides the Application Connector with the Event metadata. 
2. The Application Connector enriches the Event with the details of its source.

    > **NOTE:** There is always one dedicated instance of the Application Connector for every instance of an external solution connected to Kyma.

3. The Application Connector makes a REST API call to the **publish-knative** component and sends the enriched Event.

> 4. **publish-knative** makes the HTTP payload compatible with Knative and sends the Event to the relevant **knative-channel** service URL which is inferred based on **source id**, **event type** and **event type version** parameters.

Author: Done.

> 5. Istio Virtual Service forwards the Event further to the **nats-dispatcher** service served by the **nats-dispatcher** Pod.

Author: Done.

> 6. The NATS dispatcher stores the Event in NATS Streaming which stores the Event details in the Persistence storage volume.

Author: Done.

> 6. The NATS dispatcher stores the Event in NATS Streaming which stores the Event details in the Persistence storage volume.

Author: correct, should be nats-dispatcher.

## Event validation 

> The **subscription-controller-knative** checks if the Namespace can receive Events from the Application. It performs the check for each Kyma Subscription created in a Namespace for a particular Event type with a version for a specific Application.

Reviewer 2: ```suggestion
**subscription-controller-knative** checks if the Namespace can receive Events from the Application. It performs the check for each Kyma Subscription created in a Namespace for a particular Event type with a version for a specific Application.
```

Author: Done.

### Validation flow

See the diagram and a step-by-step description of the Event verification process.

![Event validation process](./assets/event-validation.svg)

1. The Kyma user defines a lambda or a service.
2. The Kyma user creates a Subscription custom resource.

> 3. The **subscription-controller-knative** reads the new Subscription.

Author: Done.

> 4. The **subscription-controller-knative** reads the EventActivation CR to verify if it exists in the Namespace for a certain Application.

Author: Done.

> 5. The **subscription-controller-knative**  updates the Subscription resource accordingly with the activation status `true` or `false`. The Event Bus uses this status to allow or prohibit Event delivery.

Author: Done.

title: Architecture
---

> When you create a lambda or a service to perform a given business functionality, you must define which Events trigger it. Define triggers by creating the Subscription CR in which you instruct the Event Bus to forward the Events of a particular type to your lambda. 

Reviewer: We could possibly move this inside Event consumption as the context is closer to that.

Author: Moved:)

> For example, whenever the `order-created` Event comes in, the Event Bus consumes it by saving it in NATS Streaming and persistence. Then it sends it to the receiver specified in the Subscription definition.

Reviewer: ```suggestion
For example, whenever the `order-created` Event comes in, the Event Bus stores it in NATS Streaming. The event is dispatched to the receiver specified in the Subscription definition.
```

> **NOTE:** The Event Bus creates a separate Event Trigger for each Subscription.

## Event consumption

![Configure and Consume Events](./assets/configure-consume-events.svg)

1. A user creates a lambda or a service to be triggered by an Event coming from an external solution. 
    >**NOTE**: For a service the user must create a Kyma Subscription resource manually. For a lambda, it is created automatically.

> 2. The **subscription-controller-knative** component reacts on the creation of Kyma Subscription. It [verifies](#event-validation) if the Event type has subscription permissions to if so, it creates the Knative Channel and Knative Subscription resources.

Reviewer: ```suggestion
2. The **subscription-controller-knative** component reacts on the creation of Kyma Subscription. It [verifies](#event-validation) if the Event type from the application can be consumed in the namespace in which Kyma Subscription has been created, it creates the Knative Channel and Knative Subscription resources.
```

Author: Done.

See the diagram and steps for an overview of the basic Event Bus flow:

![Event Bus architecture](./assets/event-bus-architecture.svg)

## Event flow

### Configuration

1. A user configures a lambda or a serverless application to be triggered by an Event from an external solution.

> > **NOTE**: While the system automatically creates a KymaSubscription resource for a lambda, the user must create one manually for a microservice.

Reviewer: ```suggestion
> **NOTE**: While the system automatically creates a Kyma Subscription resource for a lambda, the user must create one manually for a microservice.
```

> 2. The `subscription-controller-knative` reacts on the creation of the KymaSubscription. After verifying that the Event was activated, it creates a KnativeChannel and a KnativeSubscription.

Reviewer: ```suggestion
2. The `subscription-controller-knative` reacts on the creation of the Kyma Subscription. After verifying that the Event was activated, it creates a Knative Channel and a Knative Subscription.
```

3. The NATS controller reacts on the creation of a Knative Channel and creates the required Kubernetes and Istio services.

> 4. The NATS dispatcher reacts on the creation of a KnativeSubscription and creates the NATS Streaming subscription. 

Reviewer: ```suggestion
4. The NATS dispatcher reacts on the creation of a Knative Subscription and creates the NATS Streaming subscription. 
```

### Runtime

> 1. The external solution integrated with Kyma makes a REST API call to the Application Connector to indicate that a new Event is available.

Reviewer: we should mark this as a prerequisite.

2. The Application Connector enriches the Event with the details of its source.

> **NOTE:** There is always one dedicated instance of the Application Connector for every instance of an external solution connected to Kyma.

3. The Application Connector makes a REST API call to PublishKnative and sends the enriched Event.

> 4. PublishKnative forwards the Event to the relevant knative-channel service URL which is inferred based on **source id** , **event type** and **event type version** parameters.

Reviewer: ```suggestion
4. `publish-knative` forwards the Event to the relevant Knative Channel service URL which is inferred based on **source id** , **event type** and **event type version** parameters.
```

See the diagram and a step-by-step description of the Event verification process.

![Event validation process](./assets/event-validation.svg)

1. The Kyma user defines a lambda or a service.
2. The Kyma user creates a Subscription custom resource.

> 3. The SubscriptionController reads the new Subscription.

Reviewer: ```suggestion
3. The `subscription-controller-knative` reads the new Subscription.
```

> 4. The SubscriptionController refers to the EventActivation resource to check if the Event in the Subscription is activated for the given Namespace.

Reviewer: ```suggestion
4. The `subscription-controller-knative` refers to the Event Activation resource to check if the Event Type in the Subscription is activated for the given Namespace.
```

> 5. The SubscriptionController updates the Subscription resource accordingly.

Reviewer: ```suggestion
5. The `subscription-controller-knative` updates the Subscription resource accordingly.
```
